drivers/usb/device/usb_kinetis.c

```diff
@@ -0,0 +1,929 @@
+#endif
+
+#if (CONFIG_USB_KINETIS_NUMOF_ENDPOINTS >= 4)
+u8_t __aligned(4) ep3_buf[2][CONFIG_USB_KINETIS_EP2_SIZE];
+#define EP3_BUF_ADDR_EVEN		(u32_t)((&ep3_buf[0]))
+#define EP3_BUF_ADDR_ODD		(u32_t)((&ep3_buf[1]))
+#else
+#define EP3_BUF_ADDR_EVEN		(0)
+#define EP3_BUF_ADDR_ODD		(0)
+#endif
+
+#if (CONFIG_USB_KINETIS_NUMOF_ENDPOINTS >= 5)
+u8_t __aligned(4) ep4_buf[2][CONFIG_USB_KINETIS_EP2_SIZE];
+#define EP4_BUF_ADDR_EVEN		(u32_t)((&ep4_buf[0]))
+#define EP4_BUF_ADDR_ODD		(u32_t)((&ep4_buf[1]))
+#else
+#define EP4_BUF_ADDR_EVEN		(0)
+#define EP4_BUF_ADDR_ODD		(0)
+#endif
+
+#if (CONFIG_USB_KINETIS_NUMOF_ENDPOINTS >= 6)
+u8_t __aligned(4) ep5_buf[2][CONFIG_USB_KINETIS_EP2_SIZE];
+#define EP5_BUF_ADDR_EVEN		(u32_t)((&ep5_buf[0]))
+#define EP5_BUF_ADDR_ODD		(u32_t)((&ep5_buf[1]))
+#else
+#define EP5_BUF_ADDR_EVEN		(0)
+#define EP5_BUF_ADDR_ODD		(0)
+#endif
+
+#if (CONFIG_USB_KINETIS_NUMOF_ENDPOINTS >= 7)
+u8_t __aligned(4) ep6_buf[2][CONFIG_USB_KINETIS_EP2_SIZE];
+#define EP6_BUF_ADDR_EVEN		(u32_t)((&ep6_buf[0]))
+#define EP6_BUF_ADDR_ODD		(u32_t)((&ep6_buf[1]))
+#else
+#define EP6_BUF_ADDR_EVEN		(0)
+#define EP6_BUF_ADDR_ODD		(0)
+#endif
+
+#if (CONFIG_USB_KINETIS_NUMOF_ENDPOINTS == 8)
+u8_t __aligned(4) ep7_buf[2][CONFIG_USB_KINETIS_EP2_SIZE];
+#define EP7_BUF_ADDR_EVEN		(u32_t)((&ep7_buf[0]))
+#define EP7_BUF_ADDR_ODD		(u32_t)((&ep7_buf[1]))
+#else
+#define EP7_BUF_ADDR_EVEN		(0)
+#define EP7_BUF_ADDR_ODD		(0)
+#endif
+
+struct usb_ep_ctrl_data {
+	/* TODO: use bits for enabled, data1 ... flags */
+	bool enabled;
+	bool data1_in;
+	bool data1_out;
+	u16_t mps;
+	usb_dc_ep_callback cb;
+};
+
+struct usb_device_data {
+	usb_dc_status_callback status_cb;
+	u8_t address;
+
+	u32_t toggle_bits;
+	u32_t bd_active;
+	struct usb_ep_ctrl_data ep_ctrl[NUM_OF_EP_MAX];
+	bool attached;
+};
+
+static struct usb_device_data dev_data;
+
+static void usb_kinetis_isr_handler(void);
+
+static inline u8_t get_bdt_idx(u8_t ep, u8_t odd)
+{
+	if (ep & USB_EP_DIR_IN) {
+		return ((((KINETIS_ADDR2IDX(ep)) * 4) + 2  + (odd & 1)));
+	}
+	return ((((KINETIS_ADDR2IDX(ep)) * 4) + (odd & 1)));
+}
+
+static inline u8_t get_token_pid(u8_t idx)
+{
+	return (bdt[idx].stat >> BD_TOK_PID_SHIFT) & 0x0f;
+}
+
+static inline void set_bd_busy_odd(u8_t ep)
+{
+	if (ep & USB_EP_DIR_IN) {
+		dev_data.bd_active |=  (1 << (16 + KINETIS_ADDR2IDX(ep)));
+	} else {
+		dev_data.bd_active |= (1 << KINETIS_ADDR2IDX(ep));
+	}
+}
+
+static inline void set_bd_busy_even(u8_t ep)
+{
+	if (ep & USB_EP_DIR_IN) {
+		dev_data.bd_active &=  ~(1 << (16 + KINETIS_ADDR2IDX(ep)));
+	} else {
+		dev_data.bd_active &= ~(1 << KINETIS_ADDR2IDX(ep));
+	}
+}
+
+static inline u8_t is_bd_busy_odd(u8_t ep)
+{
+	if (ep & USB_EP_DIR_IN) {
+		return (u8_t)((dev_data.bd_active >>
+			      (16 + KINETIS_ADDR2IDX(ep))) & 1);
+	}
+	return (u8_t)((dev_data.bd_active >> KINETIS_ADDR2IDX(ep)) & 1);
+}
+
+static inline void kinetis_usb_init_clock(void)
+{
+	/* TODO: refactor or use ext/mcux sdk */
```

Comment:
Done, I wanted to make the clock source configurable, but the IRC48 did not work for me, so for the first the PLL is the only possible clock source.